README.md
# Timecard-py <!-- omit in toc -->
A simple timecard command line python script. Written by a developer, for developers. It can be configured to automatically run when you open a terminal, making this ideal for our terminal-heavy workspace. Making sure you never forget to clock in.

# Table of Contents <!-- omit in toc -->

- [Features](#features)
- [Command Usage](#command-usage)
- [Installing](#installing)
  - [Download](#download)
  - [Requirements](#requirements)
  - [Automatic Install](#automatic-install)
  - [Manual Install - Linux](#manual-install---linux)
- [Uninstalling](#uninstalling)
  - [Automatic Uninstall](#automatic-uninstall)
  - [Manual Uninstall - Linux](#manual-uninstall---linux)
- [Data Files](#data-files)
- [License](#license)

# Features

- Clock in and out at any time of the day.
- Reports how many hours you've been working.
- Coveniently gives hours worked rounded to nearest quarter.
- Reports how long you've been on break.
- Estimates when you'll be done with work, presuming an 8 hour work day and no breaks.
- Locks up terminal until you have clocked in for the first time that day (only if installed).
- Automatic cleanup. Ensures your drive won't get filled with useless timecard data.
- i3status text. Run `timecard i3status` and it will return how many hours worked to the nearest quarter, or how long you've been on break, depending on if you are clocked in or not.
- Automatic update checking (requires `requests` to be installed)

# Command Usage

`timecard [command]`
- `<no args>`: Shows time log, how many hours worked, how much time left until you're done, and how many hours you've been on break.
- `install`: Installs timecard to the user folder. See [Automatic Install](#automatic-install) for exactly what this does.
- `uninstall`: Removes timecard from the system.
- `in [offset]` (alias `i`): Clocks in if you haven't already.
  - If offset is supplied as an integer, it will clock in `[offset]` minutes ago.
  - If offset is supplied as a 24-hour time (e.g. 17:31), it will clock in at that time.
  - If no offset is supplied, it will clock in now.
- `out [offset]` (alias `o`): Clocks out if you haven't already.
  - See `in` for offset usage.
- `clock [offset]` (alias `c`): Clocks in or out, depending on current timecard state.
  - See `in` for offset usage.
- `undo` (alias `u`): Undos the last clock in/out action.
- `update`: Updates timecard to the latest version
- `version` (alias `v`): Prints the current version of timecard.
- `help` or `?`: Prints out a help message explaining command usage

# Installing

## Download

There are two different types of Timecards - the raw .py script, and built binaries. If you have Python 3.x installed, use the raw .py script. If you do not have Python 3.x for any reason, go with the built binaries as they do not require Python to be installed.  
You can download Timecard in the [releases](https://github.com/Stephen-Hamilton-C/timecard-py/releases/latest) section.

## Requirements
- Python 3 or later
- (optional) requests installed via pip3. Used for automatic updates.

## Automatic Install
- Run `timecard install`. This will perform the following:
    1. Move `timecard` to `~/.local/bin/` and remove the extension
    2. Add a line to .bashrc that will run `timecard auto`
    3. Attempt to give `timecard` executable permissions
        - If this fails, timecard will alert you and prompt if you want to add an alias instead
    4. Check if you already have `~/.local/bin` in your PATH. If not, it will prompt if you want to add it.
        - If you answer no, timecard will prompt if you want to add an alias instead
    5. Run `sudo pip3 install requests` for automatic updates
- Now you can run `timecard`

## Manual Install - Linux
1. Place `timecard` somewhere memorable.
2. Run `chmod u+x timecard` so it can be executed.
3. Add the script's parent folder to your PATH. Alternatively, you could run `sudo ln -s /path/to/timecard /usr/bin/timecard` so you don't have to update your PATH.
    - If you do update your PATH, I recommend removing the extension so you don't have to type that out every time.
    - One more alternative is making an alias to the script in your .bashrc: `alias timecard="python3 /path/to/timecard"`. You wouldn't need to change file permissions or modify your PATH this way.
4. Run `timecard`.

# Uninstalling

## Automatic Uninstall
1. Run `timecard uninstall`.

## Manual Uninstall - Linux
1. Delete the `timecard` executable/script and any symlinks you may have made.
2. Run `rm -rfv ~/.local/share/timecard/`.
3. Edit your .bashrc (`nano ~/.bashrc`) and find and remove the line under the `# Timecard autorun` comment.

# Data Files
Timecard creates timecard.json files to keep track of your clocked time. In Linux, these can be found in `~/.local/share/timecard/`. In Windows, they're in `%LocalAppData%\timecard\`.
Timecard automatically cleans out any timecard.json files that are a week old each time it is run. It will report when a timecard.json file is deleted.
The timecard.json files are stored in UNIX Epoch seconds. Their structure is as so:
```json
[
    {
        "startTime": 1234567890,
        "endTime": 1234567891
    },
    {
        "startTime": 1642739286,
        "endTime": 0
    }
]
```
If the user has not clocked out yet, the last endTime is always 0.

# License
Timecard-py is licensed under the GNU General Public License v3.0. You can find the license details in the [LICENSE](https://github.com/Stephen-Hamilton-C/timecard-py/blob/main/LICENSE) file in the main branch. Feel free to repurpose, redistribute, and branch off this code anytime. Just keep your modifications open source and licensed the same way :D

CHANGELOG.md
# Changelog <!-- omit in toc -->
All notable changes to timecard-py will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
timecard-py uses [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

# Versions <!-- omit in toc -->

- [1.1.3 - 2022-05-09](#113---2022-05-09)
- [1.1.2 - 2022-04-25](#112---2022-04-25)
- [1.1.1 - 2022-03-28](#111---2022-03-28)
- [1.1.0 - 2022-03-10](#110---2022-03-10)
- [1.0.1 - 2022-02-28](#101---2022-02-28)
- [1.0.0 - 2022-02-26](#100---2022-02-26)

# 1.1.3 - 2022-05-09

## Changed <!-- omit in toc -->
- Repo name from timecard to timecard-py

## Fixed <!-- omit in toc -->
- Incorrect help message for update command

# 1.1.2 - 2022-04-25

## Fixed <!-- omit in toc -->
- "Clock in for the day?" prompt showing when explicitly running `timecard in` for the first time.
- Updater not finding the Linux binary for the built version of timecard.

## Removed <!-- omit in toc -->
- Built versions of timecard. Only the python script is available now.

# 1.1.1 - 2022-03-28

## Fixed <!-- omit in toc -->
- Clocking in/out with negative minute offset logging as time in the future.

# 1.1.0 - 2022-03-10

## Added <!-- omit in toc -->
- Updating function. Run `timecard update` to automatically install latest update

## Changed <!-- omit in toc -->
- Update notification format
- Timecard now references itself as timecard rather than timecard.py
- Built versions of Timecard are now included

## Removed <!-- omit in toc -->
- Update checking in version command



# 1.0.1 - 2022-02-28

## Fixed <!-- omit in toc -->
- Version command causing errors due to update checking



# 1.0.0 - 2022-02-26

## Features <!-- omit in toc -->
- Clock in and out at any time of the day.
- Reports how many hours you've been working.
- Coveniently gives hours worked rounded to nearest quarter.
- Reports how long you've been on break.
- Estimates when you'll be done with work, presuming an 8 hour work day and no breaks.
- Locks up terminal until you have clocked in for the first time that day (only if installed).
- Automatic cleanup. Ensures your drive won't get filled with useless timecard data.
- i3status text. Run `timecard i3status` and it will return how many hours worked to the nearest quarter, or how long you've been on break, depending on if you are clocked in or not.
- Automatic update checking (requires `requests` to be installed)

timecard.py
#!/usr/bin/python3

###########################################################################################
# Stephen-Hamilton-C - Licensed under the GNU GPL v3 License.
# Source code can be found at https://github.com/Stephen-Hamilton-C/timecard-py
# Run `python3 timecard install` to automagically install this script into your system.
# Also install requests with `sudo pip3 install requests` for automatic update checking.

# If you fork this project, change the below constant to `<username>/<repo>`.
# If you don't, update checking and getting will come from my repo.
GITHUB_REPO = 'Stephen-Hamilton-C/timecard-py'
###########################################################################################

import sys, os, stat, json, time
from datetime import date, datetime
from platform import system
from shutil import move

class Version:
	def __init__(self, versionStr) -> None:
		version = versionStr.split('.')
		self.major = int(version[0])
		self.minor = int(version[1])
		self.patch = int(version[2])
		self.number = self.major*100 + self.minor*10 + self.patch

	def __str__(self) -> str:
		return str(self.major)+'.'+str(self.minor)+'.'+str(self.patch)

# Setup constants
VERSION = Version('1.1.3')
SCRIPT_PATH = os.path.realpath(__file__)
EXPECTED_WORK_HOURS: int = 8 * 60 * 60
TIMECARD_FILE: str = 'timecard.' + str(date.today()) + '.json'
TIMECARD_PATH: str = os.path.expanduser('~/.local/share/timecard')
INSTALL_DIR: str = os.path.expanduser('~/.local/bin')
if system() == 'Windows':
	TIMECARD_PATH = os.path.expanduser('~\\AppData\\Local\\timecard')
	INSTALL_DIR = os.path.expanduser('~')

# Constants for testing
GITHUB_BRANCH = 'main'
# TIMECARD_PATH: str = os.path.dirname(SCRIPT_PATH)

# Ensure cwd is timecard dir
if not os.path.exists(TIMECARD_PATH):
	os.makedirs(TIMECARD_PATH)
os.chdir(TIMECARD_PATH)

latestVersion = None
isInstalled = os.path.exists(os.path.join(INSTALL_DIR, 'timecard.py')) or os.path.exists(os.path.join(INSTALL_DIR, 'timecard'))
timeEntries: list = [ ]

def readFile():
	global timeEntries
	with open(TIMECARD_FILE, 'r') as timecardFile:
		timecardData: str = timecardFile.read()
		timeEntries = json.loads(timecardData)

def saveFile():
	with open(TIMECARD_FILE, 'w') as timecardFile:
		timeEntriesJson: str = json.dumps(timeEntries)
		timecardFile.write(timeEntriesJson)

def clockIn(clockedTime):
    newEntry: dict[str, int] = { "startTime": round(clockedTime), "endTime": 0 }
    timeEntries.append(newEntry)

def clockOut(clockedTime):
    lastEntry: dict[str, int] = timeEntries[-1]
    lastEntry['endTime'] = round(clockedTime)

def getClockState() -> str:
    # Clock state needs data to be loaded
	if len(timeEntries) == 0:
		try:
			readFile()
		except FileNotFoundError:
			return 'IN'
	if timeEntries[-1]['endTime'] == 0:
		return 'OUT'
	return 'IN'

def getTotalTimeWorked() -> int:
	timeSum: int = 0
	for timeEntry in timeEntries:
		if timeEntry['endTime'] == 0:
			# Use current time if this entry isn't finished yet
			timeSum += round(time.time()) - timeEntry['startTime']
		else:
			timeSum += timeEntry['endTime'] - timeEntry['startTime']
	return timeSum

def getTotalBreakTime() -> int:
	timeSum: int = 0
	for i in range(len(timeEntries)):
		timeEntry = timeEntries[i]
		if timeEntry['endTime'] == 0:
			continue
		if i+1 < len(timeEntries):
			timeSum += timeEntries[i+1]['startTime'] - timeEntry['endTime']
		else:
			timeSum += round(time.time()) - timeEntry['endTime']
	return timeSum

def getFormatter(timestamp: int) -> str:
	if timestamp < 60*60:
		# If the total time isn't even an hour, don't bother showing it
		return '%M min'
	return '%H hrs, %M min'

def remainingTimeCommand():
	# Calculate remaining time and expected end time
	totalWorked = getTotalTimeWorked()
	workRemaining = EXPECTED_WORK_HOURS - totalWorked
	endTime = round(time.time()) + workRemaining
	localized = datetime.fromtimestamp(endTime)

	formattedRemaining = time.strftime(getFormatter(workRemaining), time.gmtime(workRemaining))
	formattedLocalized = localized.strftime('%H:%M')

	print('Time remaining to get ' + str(int(EXPECTED_WORK_HOURS/60/60)) + ' hours:')
	print(formattedRemaining + ' left')
	print(formattedLocalized)
	print()


def clockCommand():
	clockState = getClockState()
	clockTime = time.time()
	if getArgument(2) != ' ':
		try:
			# Try parsing as just minutes
			offset = abs(int(getArgument(2)))
			clockTime -= offset*60
		except ValueError:
			try:
				# Try parsing as 24-hour time
				offsetTime = datetime.strptime(getArgument(2), '%H:%M')
				offsetDate = datetime.now().replace(hour=offsetTime.hour, minute=offsetTime.minute, second=0)
				offsetTimestamp = offsetDate.timestamp()

				# Sanity check
				if offsetTimestamp > clockTime:
					print('Offset cannot be after current time! Use `timecard help` to see usage of this command.')
					return

				offset = clockTime - offsetTimestamp
				clockTime -= offset
			except ValueError:
				print('Offset could not be parsed! Use `timecard help` to see usage of this command.')
				return
	if clockState == 'IN':
		# Sanity check
		if len(timeEntries) > 0 and clockTime < timeEntries[-1]['endTime']:
			print('Offset cannot be before last clock out time! Use `timecard help` to see usage of this command.')
			return
		# Check that we aren't already clocked in
		if len(timeEntries) > 0 and timeEntries[-1]['endTime'] == 0:
			print('Already clocked in! It seems another instance of timecard was running...')
			return

		clockIn(clockTime)
		remainingTimeCommand()
		if len(timeEntries) > 0:
			totalBreakTimeCommand()
	elif clockState == 'OUT':
		# Sanity check
		if clockTime < timeEntries[-1]['startTime']:
			print('Offset cannot be before last clock in time! Use `timecard help` to see usage of this command.')
			return
		# Check that we aren't already clocked out
		if timeEntries[-1]['endTime'] != 0:
			print('Already clocked out! It seems another instance of timecard was running...')
			return

		clockOut(clockTime)
		hoursWorkedCommand()
	print('Clocked ' + clockState.lower() + ' at ' + datetime.fromtimestamp(clockTime).strftime('%H:%M'))
	saveFile()

def getNearestQuarterHour(totalTime):
	return totalTime.tm_hour + (round(totalTime.tm_min/15) * 15) / 60

def hoursWorkedCommand():
	timeSum: int = getTotalTimeWorked()

	# Calculate the nearest quarter hour to input into a timesheet database
	totalTime = time.gmtime(timeSum)
	nearestQuarterHour = getNearestQuarterHour(totalTime)

	# Format the time
	formattedTime = time.strftime(getFormatter(timeSum), totalTime)

	# Report to the user the results
	print('Total time worked:')
	print(formattedTime)
	print(str(nearestQuarterHour) + ' hours')
	print()

def totalBreakTimeCommand():
	timeSum: int = getTotalBreakTime()
	totalTime = time.gmtime(timeSum)

	formattedTime = time.strftime(getFormatter(timeSum), totalTime)

	print('Total time on break:')
	print(formattedTime)
	print()

def statusCommand():
	for i in range(len(timeEntries)):
		timeEntry = timeEntries[i]
		formattedStartTime = datetime.fromtimestamp(timeEntry['startTime']).strftime('%H:%M')
		formattedEndTime = datetime.fromtimestamp(timeEntry['endTime']).strftime('%H:%M')
		if i == 0:
			print('Started work at '+formattedStartTime)
		else:
			print('Clocked in: '+formattedStartTime)

		if(timeEntry['endTime'] != 0):
			print('Clocked out: '+formattedEndTime)
		else:
			timeSum = time.time() - timeEntry['startTime']
			totalTime = time.gmtime(timeSum)
			print('Time since last clocked in: '+time.strftime(getFormatter(timeSum), totalTime))
	print()
	hoursWorkedCommand()
	remainingTimeCommand()
	totalBreakTimeCommand()

def aliasPrompt(bashrcPath, exePath):
	aliasPrompt = input('Do you want to add alias timecard="python3 '+exePath+'" to your .bashrc to get around this? (Y/n): ').strip().upper()
	if aliasPrompt != 'n':
		with open(bashrcPath, 'a') as bashrcFile:
			bashrcFile.write('alias timecard="python3 '+exePath+'" #timecard\n')

def installCommand():
	if isInstalled:
		print('timecard is already installed!')
	else:
		if not os.path.exists(INSTALL_DIR):
			os.makedirs(INSTALL_DIR)

		if system() == 'Windows':
			move(SCRIPT_PATH, os.path.join(INSTALL_DIR, __file__))
			print('Installed timecard v'+str(VERSION)+' to ' + INSTALL_DIR + '. Use `python3 '+__file__+'` to run the script')
		else:
			bashrcPath = os.path.expanduser('~/.bashrc')
			bashrcFile = open(bashrcPath, 'a')
			exePath = os.path.join(INSTALL_DIR, 'timecard')

			# Move timecard to install dir
			move(SCRIPT_PATH, exePath)

			# Append to bashrc
			bashrcFile.write('\n# Timecard autorun\n')
			bashrcFile.write('python3 ' + exePath + ' auto #timecard\n')
			try:
				# Set timecard to rwx by user
				os.chmod(exePath, stat.S_IRWXU)

				# Set PATH to include .local/bin, if not already
				PATH = os.getenv('PATH')
				if PATH.find('.local/bin') == -1:
					pathPrompt = input('Do you want to add ~/.local/bin to your PATH? This will allow you to run timecard like a command. (Y/n): ').strip().upper()
					if pathPrompt != 'n':
						bashrcFile.write('PATH=$PATH:~/.local/bin #timecard\n')
					else:
						# If user doesn't want to set path, we can try an alias
						aliasPrompt(bashrcPath, exePath)
			except Exception:
				# Probably couldn't set timecard to be rwx. We could use an alias tho
				print('Unable to make timecard executable! You\'ll have to use `python3 timecard` to run Timecard.')
				aliasPrompt(bashrcPath, exePath)
			bashrcFile.write('\n')
			print('Installed timecard v'+str(VERSION)+' to ' + INSTALL_DIR + '. Ensure that is in your PATH and then use `timecard` to run the script')
			try:
				import requests
			except ImportError:
				sudo = 'sudo '
				if system() == 'Windows':
					sudo = ''
				print('Also run `'+sudo+'pip3 install requests` for automatic update checking.')
			bashrcFile.close()


def uninstallCommand():
	if isInstalled:
		for timeFile in os.listdir():
			if timeFile.startswith('timecard.') and timeFile.endswith('.json'):
				os.remove(timeFile)
		if system() != 'Windows':
			os.chdir('..')
			os.rmdir('timecard')

			bashrcLines = []
			with open(os.path.expanduser('~/.bashrc'), 'r') as bashrcFile:
				bashrcLines = bashrcFile.readlines()
			with open(os.path.expanduser('~/.bashrc'), 'w') as bashrcFile:
				for line in bashrcLines:
					strippedLine = line.strip()
					if strippedLine != '# Timecard autorun' and not strippedLine.endswith('#timecard'):
						bashrcFile.write(line)
		print('timecard v'+str(VERSION)+' has been uninstalled!')
		os.remove(SCRIPT_PATH)
	else:
		print('timecard is not installed!')

def undoCommand():
	clockState = getClockState()
	if clockState == 'IN':
		timeEntries[-1]['endTime'] = 0
		print('Undo clock out. Currently clocked in.')
	elif clockState == 'OUT':
		timeEntries.pop(-1)
		print('Undo clock in. Currently clocked out.')

	if len(timeEntries) > 0:
		saveFile()
	elif os.path.exists(TIMECARD_FILE):
		os.remove(TIMECARD_FILE)

def getArgument(argIndex = 1) -> str:
	if len(sys.argv) > argIndex:
		return sys.argv[argIndex].strip().upper()
	return ' '

def printVersion():
	print('timecard version ' + str(VERSION))

def updateCommand():
	if checkForUpdates(False):
		import requests
		global latestVersion

		# Go to script directory
		os.chdir(os.path.dirname(SCRIPT_PATH))

		# Download new version and write to a .new file
		print('Downloading timecard.py...')
		updateRequest = requests.get('https://github.com/'+GITHUB_REPO+'/releases/download/v'+str(latestVersion)+'/timecard.py')
		updateRequest.raise_for_status()
		open('timecard.py.new', 'wb').write(updateRequest.content)

		# Replace current file with new file
		print('Updating timecard to v'+str(latestVersion)+'...')
		origName = __file__
		os.rename(__file__, 'timecard.py.old')
		os.rename('timecard.py.new', origName)

		try:
			# Set timecard to rwx by user
			os.chmod(origName, stat.S_IRWXU)
		except Exception:
			print('WARNING: Could not mark updated file as executable! Use `chmod u+x '+os.path.realpath('timecard.py')+'` to mark as executable')

		# Notify the user and delete this file
		print()
		print('Timecard has been updated!')
		print()
		os.remove('timecard.py.old')
	else:
		print('No updates available')

def checkForUpdates(alertUser = True):
	global latestVersion
	# Try to import requests
	try:
		import requests
		try:
			# Try to get latest version
			versionRequest = requests.get('https://raw.githubusercontent.com/'+GITHUB_REPO+'/'+GITHUB_BRANCH+'/version.txt')
			versionRequest.raise_for_status() # Throw a RequestException if file is not found (maybe not online)
			latestVersion = Version(versionRequest.text)

			# Notify user of new update if one is available
			if latestVersion.number > VERSION.number:
				if alertUser:
					print('----------------------------------------------------------------')
					print('An update is available for timecard!')
					print('Current version: '+str(VERSION)+', new version: '+str(latestVersion)+'.')
					print('Go to https://github.com/'+GITHUB_REPO+'/releases/latest or run `timecard update` to download the update.')
					print('----------------------------------------------------------------')
					print()
				return True
		except requests.exceptions.RequestException:
			pass
	except ImportError:
		# Requests is not installed, notify user
		sudo = 'sudo '
		if system() == 'Windows':
			sudo = ''
		print('Timecard: Unable to check for updates! To get automatic updates, run `'+sudo+'pip3 install requests`')
	return False

def printUsage():
	print('\ntimecard commands:')
	print('	<no command> - Shows time log, how many hours worked, how much time you have left to meet your desired hours worked ('+str(EXPECTED_WORK_HOURS / 60 / 60)+' hours), and how many hours you\'ve been on break.')
	print('	Install - Installs timecard to the user folder, adds an autorun to .bashrc, and adds ~/.local/bin to PATH if necessary.')
	print('	Uninstall - Removes timecard from system.')
	print('	IN (I) [offset] - Clocks in if you aren\'t already. If an offset is supplied, it logs you as clocked in OFFSET minutes ago or at OFFSET time. Time must be formatted in 24-hour time. (e.g. 17:31)')
	print('	OUT (O) [offset] - Clocks out if you aren\'t already. If an offset is supplied, it logs you as clocked out OFFSET minutes ago or at OFFSET time. Time must be formatted in 24-hour time. (e.g. 17:31)')
	print('	CLOCK (C) [offset] - Automatically determines whether to clock in/out. See IN and OUT commands.')
	print('	UNDO (U) - Undos the last clock in/out action')
	print('	Update - Updates timecard to latest version if one is available.')
	print('	Version (V) - Prints the current version of timecard.')
	print('	Help | ? - Prints this help message.')
	print()

def isCommandOrAlias(action, cmd):
	return action == cmd or (len(action) == 1 and action[0] == cmd[0])




# Check for this first so that we aren't prompting the user about clocking in if all they want to do is install
if getArgument() == 'INSTALL':
	installCommand()
elif getArgument() == 'UNINSTALL':
	uninstallCommand()
elif getArgument() == 'I3STATUS':
	if not os.path.exists(TIMECARD_FILE):
		print('OUT')
	else:
		if getClockState() == 'IN':
			breakTime: int = getTotalBreakTime()
			totalTime = time.gmtime(breakTime)
			formattedTime = time.strftime(getFormatter(breakTime), totalTime)
			print('OUT: ' + formattedTime)
		else:
			print('IN: ' + str(getNearestQuarterHour(time.gmtime(getTotalTimeWorked()))))
elif not os.path.exists(TIMECARD_FILE):
    # Timecard doesn't exist for today yet, prompt user if they want to clock in first, if they aren't doing so already
	if not isCommandOrAlias(getArgument(), 'IN') and not isCommandOrAlias(getArgument(), 'OUT'):
		prompt = input('Clock in for the day? (Y/n): ').strip().lower()
	else:
		prompt = 'y'

	# Set prompt to first char, if there are any
	if len(prompt) > 0:
		prompt = prompt[0]

	# Default answer is YES
	# Also make sure the user hasn't done this in another instance
	if prompt != 'n' and not os.path.exists(TIMECARD_FILE):
		clockCommand()

elif getArgument() == 'AUTO':
    # Script was run automatically by .bashrc (if configured that way) and timecard already exists
	print('Timecard already present for today. If you need to clock ' + getClockState().lower() + ', run `python3 '+__file__+'`')
else:
	readFile()

	# Try to get command from argument
	action = getArgument()

	if isCommandOrAlias(action, 'CLOCK'):
		clockCommand()
	elif isCommandOrAlias(action, 'IN'):
		if getClockState() == 'OUT':
			print('You\'re already clocked in! Use `timecard out` to clock out first.')
		else:
			clockCommand()
	elif isCommandOrAlias(action, 'OUT'):
		if getClockState() == 'IN':
			print('You\'re already clocked out! Use `timecard in` to clock in first.')
		else:
			clockCommand()
	elif isCommandOrAlias(action, 'UNDO'):
		undoCommand()
	elif action == 'UPDATE':
		updateCommand()
	elif isCommandOrAlias(action, 'VERSION'):
		printVersion()
	elif action == ' ':
		# Ran with no arguments
		statusCommand()
	else:
		if action[0] != '?' and action != 'HELP':
			print('Unknown command.')
		printUsage()

if getArgument() != 'I3STATUS' and getArgument() != 'UPDATE':
	checkForUpdates()

	# Cleanup old timecards, if any
	for timeFile in os.listdir():
		if os.path.isfile(timeFile) and timeFile.startswith('timecard.') and timeFile.endswith('.json') and os.stat(timeFile).st_mtime < time.time() - 7*60*60*24:
			print('Removing old timecard: '+timeFile)
			os.remove(timeFile)